### PR TITLE
Improve extracting data from op_return + add tests

### PR DESCRIPTION
**Pull request by KonradStaniec**
Fixes: https://github.com/babylonchain/babylon/issues/694

### CODE REVIEW

File: x/btccheckpoint/types/btcutils.go [hunk 1]
```diff
@@ -204,32 +192,50 @@ func verify(tx *btcutil.Tx, merkleRoot *chainhash.Hash, intermediateNodes []byte
...
+		} else {
+			// this should be one of OP_DATAXX opcodes we drop first 2 bytes
+			opReturnData = append(opReturnData, script[2:]...)
 		}
 	}
 
-	return opReturnData
+	if opReturnCounter > 1 {
+		return nil, fmt.Errorf("transaction has more than one OP_RETURN output")
+	}
```
Comment (SebastianElvis) [inline x/btccheckpoint/types/btcutils.go:236]: Can this be replaced by a nil check on `opReturnData` inside the loop? say if `txscript.IsNullData(script)` and `len(opReturnData) == 0` then return error

**Review by SebastianElvis — APPROVED** [review]

File: x/btccheckpoint/types/btcutils.go (hunk 1 above)
Comment (gitferry) [reply x/btccheckpoint/types/btcutils.go:236]: Are you trying to say add the following change to Line 208?

```
		if !txscript.IsNullData(script) {
			// not an standard op_return, we do not care about this output
			continue
		} else if len(opReturnData) != 0 {
                         // already found a valid opReturn data
                        return fmt.Errorf("transaction has more than one OP_RETURN output")
                }
```

Yeah, I think with the above change, we can get rid of the counter. No strong opinion though

File: x/btccheckpoint/types/btcutils.go [hunk 2]
```diff
@@ -204,32 +192,50 @@ func verify(tx *btcutil.Tx, merkleRoot *chainhash.Hash, intermediateNodes []byte
...
-		// - second byte should indicate how many bytes there are in opreturn script
-		if pkScriptLen > 1 &&
-			pkScriptLen <= maxOpReturnPkScriptSize &&
-			pkScript[0] == txscript.OP_RETURN {
-
-			// if this is OP_PUSHDATA1, we need to drop first 3 bytes as those are related
+		script := output.PkScript
+
+		if !txscript.IsNullData(script) {
+			// not an standard op_return, we do not care about this output
```
Comment (gitferry) [inline x/btccheckpoint/types/btcutils.go:209]:
```suggestion
			// not a standard op_return, we do not care about this output
```

File: x/btccheckpoint/types/btcutils.go [hunk 3]
```diff
@@ -204,32 +192,50 @@ func verify(tx *btcutil.Tx, merkleRoot *chainhash.Hash, intermediateNodes []byte
 	return bytes.Equal(current[:], root)
 }
 
-func ExtractOpReturnData(tx *btcutil.Tx) []byte {
+// ExtractStandardOpReturnData extract OP_RETURN data from transaction OP_RETURN
+// output.
+// If OP_RETRUN output is not standard it will be ignored. If there is more than
```
Comment (gitferry) [inline x/btccheckpoint/types/btcutils.go:197]:
```suggestion
// If OP_RETURN output is not standard it will be ignored. If there is more than
```

**Review by gitferry — APPROVED** [review]
Great fix!

File: x/btccheckpoint/types/btcutils.go (hunk 1 above)
Comment (KonradStaniec) [reply x/btccheckpoint/types/btcutils.go:236]: hmm we cannot use directly `len(opReturnData) == 0` as op_return output with empty data ie. only `OP_RETURN` opcode is an valid and standard thing. 

Although I think we can move the counter inside the loop to fail faster.

**Pull request merged**